### ISSUE TITLE
choiceRec is not really usable from C#

### ISSUE THREAD

**Issue by Porges** [hedgehogqa/fsharp-hedgehog#171]
Because C# lacks `letrec`, it's not really possible to use `choiceRec` in the way it is intended.

e.g.

```csharp
public static Gen<Expr> Expr = 
    Gen.ChoiceRecursive(
        new [] {
            from i in Gen.Constant(Gen.Int32(Range.Linear(0, 100)))
            select Expr.Constant(i)
        }, 
        new [] {
            from l in Expr
            from r in Expr
            select l + r;
       });
```

This will blow up at runtime since the `Expr` value has not been initialized yet in the `l + r` case, so it wil be `null`.

Perhaps there could be a `choiceRec` with the following signature (giving `recs` the recursive parameter that it can use):

```fsharp
    let choiceRec (nonrecs : seq<Gen<'a>>) (recs : Gen<'a> -> seq<Gen<'a>>) : Gen<'a> =
```

(I haven't yet tried to implement this to see if it's viable.)

**Comment by moodmosaic** [comment]
Or perhaps we could align this with the Haskell version and use something like [`choice`](https://github.com/hedgehogqa/haskell-hedgehog/blob/4e134f7f42365195568376eda7288fc5b318c738/hedgehog/src/Hedgehog/Internal/Gen.hs#L1138-L1150):

```haskell
-- | Randomly selects one of the generators in the list.
--
--   This generator shrinks towards the first generator in the list.
--
--   /The input list must be non-empty./
--
choice :: MonadGen m => [m a] -> m a
choice = \case
  [] ->
    error "Hedgehog.Gen.choice: used with empty list"
  xs -> do
    n <- integral $ Range.constant 0 (length xs - 1)
    xs !! n
```

**Comment by Porges** [comment]
A workaround is to wrap the recursive calls in `Gen.Delay` (although the usage is difficult since it takes `FSharpFunc`). An example is here: https://github.com/Porges/Fastre/blob/master/Fastre.Tests/Generators.cs#L19

**Comment by moodmosaic** [comment]
>although the usage is difficult since it takes `FSharpFunc`

So perhaps we could make a module that wraps and exposes all those functions in a C#-friendly way?

**Comment by Porges** [comment]
Will discuss it on #172

**Comment by adam-becker** [comment]
@moodmosaic since we closed #172 with #244, should we close this one as well? The workaround (`Gen.delay`) should now be fully usable from C#.

**Comment by moodmosaic** [comment]
Sure thing
